### DISCUSSION TITLE
csrf tokens and web security

### DISCUSSION THREAD

**Discussion by bily-yen** [community/community#141761] [Programming Help]
### Body

Hello guys,My name is billy and this is my first attempt to network with brillian people in this space.Guys i have worked with csrf tokens for the project flask-web-app-tutorial,but i get the error csrf session token is missing when i try logging in with device ip,with localhost its just works fine

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by tpmst** [comment]
Hello @bily-yen,

This problem typically arises due to how Flask handles sessions and cookies, especially when accessed via different domains or IP addresses. Here's why:

- Session Cookies: Flask uses cookies to manage user sessions. These cookies are domain-specific, meaning they are tied to the domain or IP 
address used to access the application.

- CSRF Tokens: Flask-WTF (Flask's integration with WTForms) uses CSRF tokens to protect against Cross-Site Request Forgery attacks. The CSRF token is stored in the user's session.

- Domain Differences: When you access your app via localhost, the browser correctly sends the session cookie. But when you access it via your device's IP address, the browser treats it as a different domain and may not send or accept the session cookie, leading to the missing CSRF token error.

## Solution

### 1. Ensure Secret Key is Set
The SECRET_KEY is crucial for session management and CSRF protection.
```
app = Flask(__name__)
app.config['SECRET_KEY'] = 'your_secret_key_here'
```
Replace 'your_secret_key_here' with a secure, random string.
Ensure that the secret key is set before initializing any extensions like Flask-WTF.

### 2. Configure Session Cookie Domain
Set the SESSION_COOKIE_DOMAIN configuration to include your IP address.
```
app.config['SESSION_COOKIE_DOMAIN'] = '192.168.x.x'  # Replace with your actual IP
```
This tells Flask to set the session cookie for the specified domain.

### 3. Use SERVER_NAME Configuration
Setting SERVER_NAME helps Flask understand what domain it's running on.
```
app.config['SERVER_NAME'] = '192.168.x.x:5000'  # Replace with your IP and port
```
Note: Setting SERVER_NAME can affect URL generation and routing, so use it carefully.

### 4. Update CSRF Trusted Origins
If you're using Flask-WTF version 0.15 or newer, you can specify trusted origins for CSRF protection.
```
app.config['WTF_CSRF_TRUSTED_ORIGINS'] = ['http://192.168.x.x:5000']
```
This allows CSRF tokens to be accepted from the specified IP address.

**Comment by bily-yen** [comment]
Thank you so much,i am humbled to communicate with you despite the
constraints of geographical and demographic differences, Yes,please further
scrutinize my app title the flask-web-app-tutorial and see why it is not
working for device ip incase i am unable to solve it,as a junior
developer,it is motivating to work with new people of profound experience

**Comment by bily-yen** [comment]
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from flask_login import LoginManager
from flask_socketio import SocketIO  # Import SocketIO
import pymysql
import urllib.parse
import os
from flask_limiter import Limiter
from flask_limiter.util import get_remote_address
from logging.handlers import RotatingFileHandler
import logging
import locale
from flask_cors import CORS
from flask_wtf import CSRFProtect
from flask_migrate import Migrate

# Initialize extensions globally
db = SQLAlchemy()
login_manager = LoginManager()
csrf = CSRFProtect()
limiter = Limiter(
    key_func=get_remote_address,
    storage_uri='redis://localhost:6379/0'
)
socketio = SocketIO()  # Initialize SocketIO

def create_app():
    app = Flask(__name__)

    # Load configurations
    DB_USER = os.environ.get('DB_USER', 'root')
    DB_PASSWORD = urllib.parse.quote(os.environ.get('MYSQL_PASSWORD',
'password'))
    DB_HOST = os.environ.get('DB_HOST', 'localhost')
    DB_NAME = os.environ.get('DB_NAME', 'credentials')
    DB_NAME_TONERS = os.environ.get('DB_NAME_TONERS', 'toners')
    SECRET_KEY = os.environ.get('FLASK_SECRET_KEY',
'your_default_secret_key')
    UPLOAD_FOLDER = r'static/SPAPHOTOS'  # Use a relative path for the
upload folder

    app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
    app.config['MAX_CONTENT_LENGTH'] = 16 * 1024 * 1024

    ALLOWED_EXTENSIONS = {'png', 'jpg', 'jpeg', 'gif', 'webp'}

    def allowed_file(filename):
        return '.' in filename and filename.rsplit('.', 1)[1].lower() in
ALLOWED_EXTENSIONS

    app.config.update(
        SECRET_KEY=SECRET_KEY,
        DEBUG=os.environ.get('FLASK_DEBUG', 'false').lower() in ['true', '1',
't', 'y', 'yes'],
        SQLALCHEMY_DATABASE_URI=f'mysql+pymysql://{DB_USER}:{DB_PASSWORD}@{
DB_HOST}/{DB_NAME}',
        SQLALCHEMY_BINDS={
            'toners': f'mysql+pymysql://{DB_USER}:{DB_PASSWORD}@{DB_HOST}/{
DB_NAME_TONERS}'
        },
        SQLALCHEMY_TRACK_MODIFICATIONS=False,
        SESSION_COOKIE_SECURE=not app.debug,
        SESSION_COOKIE_HTTPONLY=True,
        SESSION_COOKIE_SAMESITE='Lax',  # Use 'Lax' for a wider range of
scenarios
        SESSION_COOKIE_DOMAIN='10.0.6.144',  # Set the cookie domain
        CSRF_ENABLED=True,
        CSRF_COOKIE_SECURE=not app.debug,
        CSRF_COOKIE_SAMESITE='Lax',  # Use 'Lax' for a wider range of
scenarios
        CSRF_COOKIE_DOMAIN='10.0.6.144',  # Set the CSRF cookie domain
        SERVER_NAME='10.0.6.144:5001',  # Set the server name for routing
        WTF_CSRF_TRUSTED_ORIGINS=['http://10.0.6.144:5001'],  # Allow CSRF
tokens from this origin
    )

    # Set up logging
    handler = logging.StreamHandler()
    handler.setFormatter(logging.Formatter('%(asctime)s - %(name)s -
%(levelname)s - %(message)s'))

    if not app.debug:
        file_handler = RotatingFileHandler('app.log', maxBytes=100000,
backupCount=1)
        file_handler.setLevel(logging.INFO)
        file_handler.setFormatter(logging.Formatter('%(asctime)s - %(name)s
- %(levelname)s - %(message)s'))
        app.logger.addHandler(file_handler)
    else:
        handler.setLevel(logging.DEBUG)
        app.logger.addHandler(handler)

    app.logger.setLevel(logging.DEBUG)

    # Initialize extensions
    db.init_app(app)
    migrate = Migrate(app, db)
    csrf.init_app(app)
    login_manager.login_view = 'auth.login'
    login_manager.init_app(app)
    limiter.init_app(app)

    # Initialize SocketIO with the Flask app
    socketio.init_app(app)

    # Enable CORS with credentials support
    CORS(app, supports_credentials=True)

    # Import blueprints and register them
    from .views import views
    from .auth import auth
    app.register_blueprint(views, url_prefix='/')
    app.register_blueprint(auth, url_prefix='/')

    # Import models and create database tables
    from .models import User, Note, LoanRecord

    with app.app_context():
        db.create_all()

    # User loader for Flask-Login
    @login_manager.user_loader
    def load_user(user_id):
        return User.query.get(int(user_id))

    # Set locale for currency formatting
    locale.setlocale(locale.LC_ALL, '')

    def currency(value):
        try:
            return locale.currency(value, grouping=True)
        except (ValueError, TypeError):
            return value

    # Register the custom currency filter
    app.jinja_env.filters['currency'] = currency

    return app, socketio  # Return both app and socketio



with this still no progress

**Comment by bily-yen** [comment]
Guys help me write a financial implications for a system,i have never
really written before and i got this rare opportunity

**Comment by wasimtairq23** [comment]
Hi Billy! It sounds like you're running into an issue where CSRF tokens aren't being properly generated or recognized when logging in from your device's IP. When using localhost, the session tokens are probably managed smoothly, but switching to your device's IP might require proper configuration of the session cookies. Make sure that you're setting your CSRF tokens correctly and that the IP address is properly recognized by your Flask app. Also, check your Flask session settings, especially around cross-origin access. If you're looking to deepen your understanding, [web developer courses](https://itcourses.au/courses/web-development-training-course/) often cover topics like CSRF protection in detail.
